x-pack/plugin/search-business-rules/src/main/java/org/apache/lucene/search/CappedScoreQuery.java

```diff
@@ -0,0 +1,170 @@
+        this.query = Objects.requireNonNull(query, "Query must not be null");
+        this.maxScore = maxScore;
+    }
+
+    /** Returns the encapsulated query. */
+    public Query getQuery() {
+        return query;
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        Query rewritten = query.rewrite(reader);
+
+        if (rewritten != query) {
+            return new CappedScoreQuery(rewritten, maxScore);
+        }
+
+        if (rewritten.getClass() == CappedScoreQuery.class) {
+            return rewritten;
+        }
+
+        if (rewritten.getClass() == BoostQuery.class) {
+            return new CappedScoreQuery(((BoostQuery) rewritten).getQuery(), maxScore);
+        }
+
+        return super.rewrite(reader);
+    }
+
+    /**
+     * We return this as our {@link BulkScorer} so that if the CSQ wraps a query with its own optimized top-level scorer (e.g.
+     * BooleanScorer) we can use that top-level scorer.
+     */
+    protected static class CappedBulkScorer extends BulkScorer {
+        final BulkScorer bulkScorer;
+        final Weight weight;
+        final float maxScore;
+
+        public CappedBulkScorer(BulkScorer bulkScorer, Weight weight, float maxScore) {
+            this.bulkScorer = bulkScorer;
+            this.weight = weight;
+            this.maxScore = maxScore;
+        }
+
+        @Override
+        public int score(LeafCollector collector, Bits acceptDocs, int min, int max) throws IOException {
+            return bulkScorer.score(wrapCollector(collector), acceptDocs, min, max);
+        }
+
+        private LeafCollector wrapCollector(LeafCollector collector) {
+            return new FilterLeafCollector(collector) {
+                @Override
+                public void setScorer(Scorable scorer) throws IOException {
+                    // we must wrap again here, but using the scorer passed in as parameter:
+                    in.setScorer(new FilterScorable(scorer) {
+                        @Override
+                        public float score() throws IOException {
+                            return Math.min(maxScore, in.score());
+                        }
+                    });
+                }
+            };
+        }
+
+        @Override
+        public long cost() {
+            return bulkScorer.cost();
+        }
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        final Weight innerWeight = searcher.createWeight(query, ScoreMode.COMPLETE, 1f);
+        if (scoreMode.needsScores()) {
+            return new CappedScoreWeight(this, innerWeight, boost) {
+                @Override
+                public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
+                    if (scoreMode == ScoreMode.TOP_SCORES) {
+                        return super.bulkScorer(context);
+                    }
```

Comment:
if we implement setMinCompetitiveScore in the above FilterScorable then we wouldn't need this special case?